include/bark/client.hpp

```diff
@@ -1,27 +1,68 @@
 #pragma once
 
 #include <cstdint>
+#include <utility>
 
 #include "bark/datagram.hpp"
+#include "bark/feature_detection.hpp"
 #include "bark/i_datadog_client.hpp"
 #include "bark/tags.hpp"
-#include "bark/udp_client.hpp"
+#include "bark/transports/async_udp_transport.hpp"
+#include "bark/transports/async_uds_transport.hpp"
+#include "bark/transports/constants.hpp"
+#include "bark/transports/datagram_transport.hpp"
+#include "bark/transports/udp_transport.hpp"
+#include "bark/transports/uds_transport.hpp"
 
 namespace bark
 {
 
+template<datagram_transport Transport>
 class Client final : public IDatadogClient
 {
-    UDPClient _udp_client;
-    Tags _global_tags;
+    Transport _transport;
 
   public:
-    explicit Client(UDPClient&& udp_client, Tags global_tags = no_tags);
+    explicit Client(Transport&& transport)
+        : _transport(std::move(transport))
+    {
+    }
 
-    auto send(const Datagram& datagram) -> void override;
-    auto send(Datagram&& datagram) -> void override;
+    auto send(const Datagram& datagram) -> void override
+    {
+        if constexpr (sync_datagram_transport<Transport>) {
+            this->_transport.send(Datagram {datagram});
+        } else {
+            this->_transport.send_async(Datagram {datagram});
+        }
+    }
 
-    static auto make_local_client(Tags global_tags = no_tags, uint16_t port = dogstatsd_udp_port) -> Client;
+    auto send(Datagram&& datagram) -> void override
+    {
+        if constexpr (sync_datagram_transport<Transport>) {
+            this->_transport.send(std::move(datagram));
+        } else {
+            this->_transport.send_async(std::move(datagram));
+        }
+    }
 };
 
+using UDPClient = Client<transports::UDPTransport>;
+
+inline auto make_local_udp_client(Tags global_tags = no_tags,
+                                  uint16_t port = transports::dogstatsd_udp_port) -> UDPClient
+{
+    return UDPClient {transports::UDPTransport::make_local_udp_transport(port, std::move(global_tags))};
+}
+
+extern template class Client<transports::AsyncUDPTransport>;
+extern template class Client<transports::UDPTransport>;
+
+#if BARK_UDS_ENABLED
```

Comment:
is the reason to have this, that windows does not support it in asio?

Reply:
Not asio specifically.
UDS = Unix Domain Socket = Not available on Windows.